src/util/apparmor.c

```diff
@@ -93,3 +105,453 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
+                                                path, interface, method, allow, audit);
+        } else if (nameset->type == NAME_SET_TYPE_OWNER) {
+                if (c_rbtree_is_empty(&nameset->owner->ownership_tree)) {
+                        struct Address addr;
+                        address_init_from_id(&addr, subject_id);
+                        r = apparmor_message_query_name(check_send, security_label, bustype,
+                                                        receiver_context, address_to_string(&addr),
+                                                        path, interface, method, allow, audit);
+                } else {
+                        /*
+                         * A set of owned names is given. In this case, we iterate all
+                         * of them and match against each. Note that this matches even
+                         * on non-primary name owners.
+                         */
+                        c_rbtree_for_each_entry(ownership, &nameset->owner->ownership_tree, owner_node) {
+                                r = apparmor_message_query_name(check_send, security_label,
+                                                                bustype, receiver_context,
+                                                                ownership->name->name, path,
+                                                                interface, method, allow, &audit_tmp);
+                                if (r < 0)
+                                        return r;
+                                if (!allow)
+                                        return r;
+                                if (audit_tmp)
+                                        *audit = 1;
+                        }
+                }
+        } else if (nameset->type == NAME_SET_TYPE_SNAPSHOT) {
+                /*
+                 * An ownership-snapshot is given. Again, we simply iterate the
+                 * names and match each. Note that the snapshot must contain
+                 * queued names as well, since the policy matches on it.
+                 */
+                for (i = 0; i < nameset->snapshot->n_names; ++i) {
+                        r = apparmor_message_query_name(check_send, security_label, bustype,
+                                                        receiver_context,
+                                                        nameset->snapshot->names[i]->name,
+                                                        path, interface, method, allow, &audit_tmp);
+                        if (r < 0)
+                                return r;
+                        if (!allow)
+                                return r;
+                        if (audit_tmp)
+                                *audit = 1;
+                }
+        } else if (nameset->type != NAME_SET_TYPE_EMPTY) {
+                c_assert(0);
+                r = -EINVAL;
+        }
+
+        return r;
```

Comment:
Uff. This is slow. I mean, we optimized everything so we have a single allocation per message transaction. Now the apparmor code adds one allocation per name the receiver and sender own. I am fine keeping the code, but man those LSMs really drag things down. Makes me a bit sad :-/